backend/src/mirrors_qa_backend/db/mirrors.py

```diff
@@ -0,0 +1,140 @@
+    current_countries: dict[str, schemas.Country] = {
+        country.code: country for country in countries
+    }
+
+    for country_code, country in current_countries.items():
+        if country_code not in db_countries:
+            # Register all of the country's mirrors as the country is
+            # a new country
+            logger.debug("Registering new mirrors for {country_code!r}")
+            c = models.Country(code=country.code, name=country.name)
+            c.mirrors = [models.Mirror(**m.model_dump()) for m in country.mirrors]
+            session.add(c)
+
+    for code, db_country in db_countries.items():
+        if code in current_countries:
+            # Even though the db_country is "current", ensure it's mirrors
+            # are in sync with the current mirrors
+            current_mirrors: dict[str, schemas.Mirror] = {
+                m.id: m for m in current_countries[code].mirrors
+            }
+            db_mirrors: dict[str, models.Mirror] = {m.id: m for m in db_country.mirrors}
+
+            for db_mirror in db_mirrors.values():
+                if db_mirror.id not in current_mirrors:
+                    logger.debug(f"Disabling mirror {db_mirror.id!r}")
+                    db_mirror.enabled = False
+                    session.add(db_mirror)
+
+            for mirror_id, mirror in current_mirrors.items():
+                if mirror_id not in db_mirrors:
+                    logger.debug(
+                        f"Registering new mirror {mirror.id!r} for "
+                        "country: {db_country.name!r}"
+                    )
+                    db_country.mirrors.append(models.Mirror(**mirror.model_dump()))
+                    session.add(db_country)
+        else:
+            # disable all of the country's mirrors as they have been removed
+            for db_mirror in db_country.mirrors:
+                logger.debug(f"Disabling mirror {db_mirror.id!r}")
+                db_mirror.enabled = False
+                session.add(db_mirror)
+
+
+def get_current_mirror_countries() -> list[schemas.Country]:
+    def find_country_rows(tag: Tag) -> bool:
+        """
+        Filters out table rows that do not contain mirror
+        data from the table body.
+        """
+        return tag.name == "tr" and tag.findChild("td", class_="newregion") is None
+
+    r = requests.get(Settings.mirrors_url, timeout=Settings.requests_timeout)
+    r.raise_for_status()
+
+    soup = BeautifulSoup(r.text, features="html.parser")
+    body = soup.find("tbody")
+
+    if body is None or isinstance(body, NavigableString):
+        raise ValueError
+    # Given a country might have more than one mirror, set up a dictionary
+    # of country_code to the country's data. If it is the first time we
+    # are seeing the country, we save it along with its mirror, else,
+    # we simply update its mirrors list.
+    countries: dict[str, schemas.Country] = {}
+    rows = body.find_all(find_country_rows)
```

Comment:
The `rows` variable represents the rows where the country information starts from while the `countries` dictionary acts as a map to compare if we have seen a country before as some countries have multiple mirrors.

Perhaps, you mean the `find_country_rows` function. I can remove that and refactor it to a lambda function if that is what you mean.

Reply:
No I meant instead of 

```py
rows = body.find_all(find_country_rows)
for row in rows:
```

we could have:

```py
for row in body.find_all(find_country_rows):
```